pkg/sync/expand.go

```diff
@@ -20,81 +21,158 @@ type EntitlementGraphAction struct {
 	PageToken               string   `json:"page_token"`
 }
 
-type edgeInfo struct {
+// Edges between entitlements are grants.
+type grantInfo struct {
 	Expanded        bool     `json:"expanded"`
 	Shallow         bool     `json:"shallow"`
 	ResourceTypeIDs []string `json:"resource_type_ids"`
 }
 
+type Node struct {
+	Id             int      `json:"id"`
+	EntitlementIDs []string `json:"entitlementIds"` // List of entitlements.
+}
+
 type EntitlementGraph struct {
-	Entitlements map[string]bool                 `json:"entitlements"`
-	Edges        map[string]map[string]*edgeInfo `json:"edges"`
-	Loaded       bool                            `json:"loaded"`
-	Depth        int                             `json:"depth"`
-	Actions      []EntitlementGraphAction        `json:"actions"`
+	NodeCount int                        `json:"node_count"`
+	Nodes     map[int]Node               `json:"nodes"`
+	Edges     map[int]map[int]*grantInfo `json:"edges"` // Adjacency list. Source node -> destination node
+	Loaded    bool                       `json:"loaded"`
+	Depth     int                        `json:"depth"`
+	Actions   []EntitlementGraphAction   `json:"actions"`
 }
 
 func NewEntitlementGraph(ctx context.Context) *EntitlementGraph {
 	return &EntitlementGraph{
-		Entitlements: make(map[string]bool),
-		Edges:        make(map[string]map[string]*edgeInfo),
+		NodeCount: 1, // Start at 1 in case we don't initialize something and try to get node 0
+		Nodes:     make(map[int]Node),
+		// TODO: probably want this for efficiency
+		// EntitlementsToNodes: make(map[string]int),
+		Edges: make(map[int]map[int]*grantInfo),
 	}
 }
 
-// IsExpanded returns true if all entitlements in the graph have been expanded.
-func (d *EntitlementGraph) IsExpanded() bool {
-	for entilementID := range d.Entitlements {
-		if !d.IsEntitlementExpanded(entilementID) {
+func (g *EntitlementGraph) Str() string {
+	str := "\n"
+	for id, node := range g.Nodes {
+		str += fmt.Sprintf("node %v entitlement IDs: %v\n", id, node.EntitlementIDs)
```

Comment:
should we be using a `strings.Builder` for this? Not sure how many iterations happen here

Reply:
Also just realized this is probably mostly used for debugging, so maybe doesn't matter.

Reply:
Yeah I only used it for debugging. It's not called anywhere now.